File: ios/improv_wifi/Sources/improv_wifi/ImprovWifiPlugin.swift
import Flutter
import UIKit
import CoreBluetooth
import Combine
#if canImport(ImproviOS)
import ImproviOS
#elseif canImport(Improv_iOS)
import Improv_iOS
#endif

public class ImprovWifiPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private var cancellables = Set<AnyCancellable>()
    private var peripheralMap: [String: CBPeripheral] = [:]

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "be.azimut.improv_wifi/methods",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "be.azimut.improv_wifi/state",
            binaryMessenger: registrar.messenger()
        )

        let instance = ImprovWifiPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startScan":
            ImprovManager.shared.scan()
            result(nil)
        case "stopScan":
            ImprovManager.shared.stopScan()
            result(nil)
        case "connectToDevice":
            guard let args = call.arguments as? [String: Any],
                  let deviceId = args["deviceId"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "deviceId is required", details: nil))
                return
            }
            guard let peripheral = peripheralMap[deviceId] else {
                result(FlutterError(code: "DEVICE_NOT_FOUND", message: "Device with id \(deviceId) not found", details: nil))
                return
            }
            ImprovManager.shared.connectToDevice(peripheral)
            result(nil)
        case "disconnectDevice":
            if let peripheral = ImprovManager.shared.connectedDevice {
                ImprovManager.shared.disconnectFromDevice(peripheral)
            }
            result(nil)
        case "identifyDevice":
            ImprovManager.shared.identifyDevice()
            result(nil)
        case "sendWifi":
            guard let args = call.arguments as? [String: Any],
                  let ssid = args["ssid"] as? String,
                  let password = args["password"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "ssid and password are required", details: nil))
                return
            }
            ImprovManager.shared.sendWifi(ssid: ssid, password: password)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - FlutterStreamHandler

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        setupSubscriptions()
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        cancellables.removeAll()
        return nil
    }

    // MARK: - Private

    private func setupSubscriptions() {
        guard let manager = ImprovManager.shared as? ImprovManager else { return }

        // Combine all @Published properties into a single stream to avoid duplicate emissions
        Publishers.CombineLatest3(
            Publishers.CombineLatest3(
                manager.$foundDevices,
                manager.$connectedDevice,
                manager.$bluetoothState
            ),
            Publishers.CombineLatest(
                manager.$deviceState,
                manager.$errorState
            ),
            manager.$lastResult
        )
        .removeDuplicates { prev, curr in
            // Compare all values to detect actual changes
            let (prevGroup1, prevGroup2, prevLastResult) = prev
            let (currGroup1, currGroup2, currLastResult) = curr
            let (prevDevices, prevConnected, prevBluetooth) = prevGroup1
            let (currDevices, currConnected, currBluetooth) = currGroup1
            let (prevDeviceState, prevErrorState) = prevGroup2
            let (currDeviceState, currErrorState) = currGroup2

            return prevDevices.keys == currDevices.keys &&
                   prevConnected?.identifier == currConnected?.identifier &&
                   prevBluetooth == currBluetooth &&
                   prevDeviceState == currDeviceState &&
                   prevErrorState == currErrorState &&
                   prevLastResult == currLastResult
        }
        .receive(on: DispatchQueue.main)
        .sink { [weak self] _ in
            self?.sendStateUpdate()
        }
        .store(in: &cancellables)
    }

    private func sendStateUpdate() {
        guard let eventSink = eventSink else { return }

        let manager = ImprovManager.shared

        // Update peripheral map (SDK uses [String: CBPeripheral])
        peripheralMap.removeAll()
        for peripheral in manager.foundDevices.values {
            peripheralMap[peripheral.identifier.uuidString] = peripheral
        }

        // Build found devices list
        let foundDevices: [[String: Any?]] = manager.foundDevices.values.map { peripheral in
            return [
                "id": peripheral.identifier.uuidString,
                "name": peripheral.name
            ]
        }

        // Build state map
        var state: [String: Any?] = [
            "foundDevices": foundDevices,
            "connectedDeviceId": manager.connectedDevice?.identifier.uuidString,
            "bluetoothState": bluetoothStateToString(manager.bluetoothState),
            "deviceState": deviceStateToString(manager.deviceState),
            "errorState": errorStateToString(manager.errorState),
            "lastResult": manager.lastResult
        ]

        eventSink(state)
    }

    private func bluetoothStateToString(_ state: CBManagerState) -> String {
        switch state {
        case .unknown: return "unknown"
        case .resetting: return "resetting"
        case .unsupported: return "unsupported"
        case .unauthorized: return "unauthorized"
        case .poweredOff: return "poweredOff"
        case .poweredOn: return "poweredOn"
        @unknown default: return "unknown"
        }
    }

    private func deviceStateToString(_ state: DeviceState?) -> String? {
        guard let state = state else { return nil }
        switch state {
        case .authorizationRequired: return "authorizationRequired"
        case .authorized: return "authorized"
        case .provisioning: return "provisioning"
        case .provisioned: return "provisioned"
        @unknown default: return nil
        }
    }

    private func errorStateToString(_ state: ErrorState?) -> String? {
        guard let state = state else { return nil }
        switch state {
        case .noError: return "noError"
        case .invalidRPCPacket: return "invalidRPCPacket"
        case .unknownCommand: return "unknownCommand"
        case .unableToConnect: return "unableToConnect"
        case .notAuthorized: return "notAuthorized"
        case .unknown: return "unknown"
        @unknown default: return "unknown"
        }
    }
}

File: android/src/main/kotlin/com/example/improv_wifi/ImprovWifiPlugin.kt
package com.example.improv_wifi

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
// SDK classes are now in the same package
// import com.wifi.improv.ImprovManager
// import com.wifi.improv.ImprovManagerCallback
// import com.wifi.improv.ImprovDevice
// import com.wifi.improv.DeviceState
// import com.wifi.improv.ErrorState

class ImprovWifiPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var context: Context? = null
    private var improvManager: ImprovManager? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    // State tracking
    private val foundDevices = mutableMapOf<String, ImprovDevice>()
    private var connectedDevice: ImprovDevice? = null
    private var isScanning = false
    private var currentDeviceState: DeviceState? = null
    private var currentErrorState: ErrorState? = null
    private var lastResult: List<String>? = null

    private val callback = object : ImprovManagerCallback {
        override fun onScanningStateChange(scanning: Boolean) {
            isScanning = scanning
            sendStateUpdate()
        }

        override fun onDeviceFound(device: ImprovDevice) {
            foundDevices[device.address] = device
            sendStateUpdate()
        }

        override fun onConnectionStateChange(device: ImprovDevice?) {
            connectedDevice = device
            if (device == null) {
                // Reset states when disconnected
                currentDeviceState = null
                currentErrorState = null
                lastResult = null
            }
            sendStateUpdate()
        }

        override fun onStateChange(state: DeviceState) {
            currentDeviceState = state
            sendStateUpdate()
        }

        override fun onErrorStateChange(errorState: ErrorState) {
            currentErrorState = errorState
            sendStateUpdate()
        }

        override fun onRpcResult(result: List<String>) {
            lastResult = result
            sendStateUpdate()
        }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext

        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.example.improv_wifi/methods")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "com.example.improv_wifi/state")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        context = null
        improvManager = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "startScan" -> {
                ensureManagerInitialized()
                foundDevices.clear()
                improvManager?.findDevices()
                result.success(null)
            }
            "stopScan" -> {
                improvManager?.stopScan()
                result.success(null)
            }
            "connectToDevice" -> {
                android.util.Log.d("ImprovWifiPlugin", "=== connectToDevice method called ===")
                val deviceId = call.argument<String>("deviceId")
                android.util.Log.d("ImprovWifiPlugin", "deviceId: $deviceId")
                if (deviceId == null) {
                    android.util.Log.e("ImprovWifiPlugin", "deviceId is null!")
                    result.error("INVALID_ARGUMENTS", "deviceId is required", null)
                    return
                }
                val device = foundDevices[deviceId]
                android.util.Log.d("ImprovWifiPlugin", "device found in foundDevices: ${device != null}")
                android.util.Log.d("ImprovWifiPlugin", "foundDevices contains: ${foundDevices.keys}")
                if (device == null) {
                    android.util.Log.e("ImprovWifiPlugin", "Device not found in foundDevices!")
                    result.error("DEVICE_NOT_FOUND", "Device with id $deviceId not found", null)
                    return
                }
                android.util.Log.i("ImprovWifiPlugin", "Calling improvManager.connectToDevice for ${device.name} (${device.address})")
                improvManager?.connectToDevice(device)
                result.success(null)
                android.util.Log.d("ImprovWifiPlugin", "connectToDevice completed successfully")
            }
            "disconnectDevice" -> {
                // Note: The Android SDK doesn't have a disconnectDevice method exposed
                // We reset the state
                connectedDevice = null
                currentDeviceState = null
                currentErrorState = null
                lastResult = null
                sendStateUpdate()
                result.success(null)
            }
            "identifyDevice" -> {
                try {
                    improvManager?.identifyDevice()
                    result.success(null)
                } catch (e: Exception) {
                    result.error("NOT_CONNECTED", e.message, null)
                }
            }
            "sendWifi" -> {
                val ssid = call.argument<String>("ssid")
                val password = call.argument<String>("password")
                if (ssid == null || password == null) {
                    result.error("INVALID_ARGUMENTS", "ssid and password are required", null)
                    return
                }
                try {
                    improvManager?.sendWifi(ssid, password)
                    result.success(null)
                } catch (e: Exception) {
                    result.error("NOT_CONNECTED", e.message, null)
                }
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        ensureManagerInitialized()
        sendStateUpdate()
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private fun ensureManagerInitialized() {
        android.util.Log.d("ImprovWifiPlugin", "ensureManagerInitialized: improvManager=${improvManager != null}, context=${context != null}")
        if (improvManager == null && context != null) {
            android.util.Log.i("ImprovWifiPlugin", "Creating new ImprovManager instance")
            improvManager = ImprovManager(context!!, callback)
            android.util.Log.i("ImprovWifiPlugin", "ImprovManager created successfully")
        }
    }

    private fun sendStateUpdate() {
        val sink = eventSink ?: return

        val devicesList = foundDevices.values.map { device ->
            mapOf(
                "id" to device.address,
                "name" to device.name
            )
        }

        val state = mapOf(
            "foundDevices" to devicesList,
            "connectedDeviceId" to connectedDevice?.address,
            "bluetoothState" to "poweredOn", // Android SDK doesn't expose BT state directly
            "deviceState" to deviceStateToString(currentDeviceState),
            "errorState" to errorStateToString(currentErrorState),
            "lastResult" to lastResult
        )

        // CRITICAL: EventChannel must be called on the main thread
        mainHandler.post {
            sink.success(state)
        }
    }

    private fun deviceStateToString(state: DeviceState?): String? {
        return when (state) {
            DeviceState.AUTHORIZATION_REQUIRED -> "authorizationRequired"
            DeviceState.AUTHORIZED -> "authorized"
            DeviceState.PROVISIONING -> "provisioning"
            DeviceState.PROVISIONED -> "provisioned"
            null -> null
        }
    }

    private fun errorStateToString(state: ErrorState?): String? {
        return when (state) {
            ErrorState.NO_ERROR -> "noError"
            ErrorState.INVALID_RPC_PACKET -> "invalidRPCPacket"
            ErrorState.UNKNOWN_COMMAND -> "unknownCommand"
            ErrorState.UNABLE_TO_CONNECT -> "unableToConnect"
            ErrorState.NOT_AUTHORIZED -> "notAuthorized"
            ErrorState.UNKNOWN -> "unknown"
            null -> null
        }
    }
}
